# 5 interesting functions

In this notebook, I have chosen 5 different functions in class tensor in Pytorch:
- new_ones() : create a tensor filled by 1
- T : transpose of a tensor
- abs_() : get a tensor with absolute values
- add_() : add up two tensors
- argmin() : find a minimum value in a tensor

In [153]:
# Import torch and other required modules
import torch

## Function 1 - torch.new_ones

Explanation:
Function 'new_ones' is used to create a matrix filled by 1 where number of rows and columns can be passed in parameters as follows:
new_ones((nbOfRows, nbOfCol)) where nbOfRows and nbOfCol are of types int.
This function is called on a tensor object.

In [25]:
# Example 1
x = torch.tensor([[1., -1., -3.], [1., -1., 5.]])
x.new_ones((2, 8))

tensor([[1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1.]])

Explanation about example 1 :
Firstly, we create a tensor object with any size (here 2*3) and any type (here float).
Secondly, we create a matrix filled by 1 of size (2, 8) and the same type data as initial tensor object.

In [26]:
# Example 2 - working
x = torch.tensor([[2, 5], [1, 2], [6, 7]])
x.new_ones((4, 1))

tensor([[1],
        [1],
        [1],
        [1]])

Explanation about example 2 :
Firstly, we create a tensor object with size (here 3*2) and type int.
Secondly, we create a matrix filled by 1 of size (4,1) and type int.

In [28]:
# Example 3 - breaking (to illustrate when it breaks)
x = torch.tensor([[2, 5, 5], [1, 2], [6, 7]])
x.new_ones((4, 1))

ValueError: expected sequence of length 3 at dim 1 (got 2)

Explanation about example 3 :
The error is caused by creating a matrix of incoherent size: the first row has 3 columns whereas the second and the third row have only two columns.

Closing comments about when to use 'new_ones': When we need to create a matrix of 1 from a tensor object, we use this function.

## Function 2 - torch.T

Explanation: 
Function 'T' is used to create a matrix transposed from the initial as follows:
Suppose that we have a tensor object 'ts' of size (nbOfRows, nbOfCol). To use this function, we call 'ts.T'.
The returned object is also a tensor whose size is of (nbOfCol, nbOfRows).

In [33]:
# Example 1 - working
y = torch.tensor([[1., -8., -3.], [2., -1., 3.]])
y.T

tensor([[ 1.,  2.],
        [-8., -1.],
        [-3.,  3.]])

Explanation about example 1 :
We have a tensor object called y whose size is of (2, 3).
Then the function T is called on y, that gives us the new tensor of size (3, 2) of the same data type as the initial one.
The new created matrix is the transpose of y, that means: y(i, j) == T(j, i), 0 <= i <= 1 and 0 <= j <= 2

In [62]:
# Example 2 - working
y = torch.tensor([[100, 150]])
y.T

tensor([[100],
        [150]])

Explanation about example 2 :
The tensor object is of size (1, 2).
The function transpose T is called on y, giving us the matrix of size (2, 1).
y(0,0) == T(0,0) and y(0,1) == T(1,0)

In [36]:
# Example 3 - breaking (to illustrate when it breaks)
y = torch.tensor([[100, 1], [2]])
y.T

ValueError: expected sequence of length 2 at dim 1 (got 1)

Explanation about example 3 :
The tensor object y is badly created, the first row has two columns, whereas the second row has only one column.
Then, the transpose T can not be done. This causes this error.

Closing comments about when to use 'T':
This function is used to create the transpose of a tensor object (matrix).

## Function 3 - torch.abs_()

Explanation: 
Function 'abs_()' is used to turn values in a matrix to absolute values (change in-place) :
Suppose that we have a tensor object 'ts'.
When we call abs_() on ts, all negative values in this matrix are replaced by its absolute values, the positive values remain unchanged.

In [59]:
# Example 1 - working
z = torch.tensor([-2.5, 1.5])
z.abs_()

tensor([2.5000, 1.5000])

Explanation about example 1:
When we call abs_() on z, all values in z are replaced by their absolute values.

In [58]:
# Example 2 - working
z = torch.tensor([-10, 5])
z.abs_()

tensor([10,  5])

Explanation about example 2:
When we call abs_() on z, -10 becomes 10 and 5 becomes 5.

In [60]:
# Example 3 - breaking (to illustrate when it breaks)
z = torch.tensor([-3, 'a'])
z.abs_()

TypeError: new(): invalid data type 'str'

Explanation about example 3:
z is a tensor of size (1, 2). The first element is of data type int -3, however the second element is of data type string, on which we can not apply the function abs_().

Closing comments about when to use 'abs_()':
This function is used when we need to ensure that all elements in a matrix are greater than 0.

## Function 4 - add_(u, *, alpha=1)

Each element of the tensor u is multiplied by the scalar alpha and added to each element of the initial tensor.

In [100]:
# Example 1 - working
t = torch.tensor([5, 3])
u = torch.tensor([3])
t.add_(u, alpha=2)

tensor([11,  9])

Explanation about example 1:
The elements in tensor resulted from this function is explained as follows:
5+3*alpha = 11 and 3+3*alpha = 9.
Hence we obtain a tensor [11, 9].

In [124]:
# Example 2 - working
t = torch.tensor([[5, 3], [1, 4]])
u = torch.tensor([[3], [1]])
t.add_(u, alpha=2)

tensor([[11,  9],
        [ 3,  6]])

Explanation about example 2:
The elements in tensor resulted from this function is explained as follows:
5+3*alpha = 11 and 3+3*alpha = 9 for the first row.
1+1*alpha = 3 and 4+1*alpha = 6 for the second row.
Hence we obtain a tensor [[11,  9], [ 3,  6]].

In [110]:
# Example 3 - breaking (to illustrate when it breaks)
t = torch.tensor([5, 3])
u = torch.tensor([3, 1, 2])
t.add_(u, alpha=2)

RuntimeError: The size of tensor a (2) must match the size of tensor b (3) at non-singleton dimension 0

Explanation about example 3:
The tensor t is of size (1, 2) and the tensor u is of size (1, 3), we cannot add u*alpha up to t as their size is incoherent.
Hence we obtain the error above.

Closing comments about when to use add_():
We use this function when we need to add quickly two tensors broadcastable.

## Function 5 - argmin

This function is used to return the indice of the minimum values in a tensor.

In [125]:
# Example 1 - working
v = torch.tensor([[1., -8., -3.], [2., -1., 3.]])
v.argmin()

tensor(1)

Explanation about example 1:
In the tensor v, the minimum value is -8, found at indice 1. We remark that the first element in the first row (i.e. 1) has indice 0.

In [130]:
# Example 2 - working
v = torch.randn(4, 4)
v

tensor([[ 0.5009, -1.6985, -1.5920, -0.8039],
        [-0.3223, -1.3781, -1.9622,  0.0600],
        [ 0.2508, -0.0455, -1.7420,  0.2262],
        [-0.0254, -0.1017, -0.4427,  1.9173]])

In [131]:
v.argmin()

tensor(6)

Explanation about example 2:
In this example, v is defined as a random matrix of size (4,4).
Its minimum element is found at the indice 6 (-1.9622), counting from (0,0), then go right to the end of first row, then go down to the first element of the second row, etc.

In [146]:
# Example 3 - breaking (to illustrate when it breaks)
v = torch.randn(5, 5)
v

tensor([[ 2.2573,  0.1159,  0.1859, -0.2176, -0.1569],
        [ 1.2720,  0.5931, -1.0577, -0.8316, -0.9798],
        [ 0.4953, -0.8367,  2.1577, -1.0104,  0.7332],
        [-0.6326,  1.0546,  1.2790, -0.7770,  0.8817],
        [ 0.4183,  0.9912,  0.6161,  0.5378, -0.6180]])

In [154]:
v.argmin(dim=2)

IndexError: Dimension out of range (expected to be in range of [-2, 1], but got 2)

Explanation about example 3:
This error is caused since we passed in parameter dim=2. The range of this value is [-2,1].

When this value is:
-2 or 0, the function returns min value in its column
-1 or 1, the function returns min value in its row

Closing comments about when to use argmin:
When we need to find out the minimum value in a tensor.
See more: argmax

## Conclusion

In this notebook, I have chosen 5 different functions in class tensor in Pytorch.
They are new_ones(), T, abs_(), add_(), argmin(), each of them has a particular functionality which are useful in the computation of tensor (matrix).

## Reference Links
Provide links to your references and other interesting articles about tensors
* Official documentation for `torch.Tensor`: https://pytorch.org/docs/stable/tensors.html
* Two broadcasting tensors: https://pytorch.org/docs/stable/notes/broadcasting.html#broadcasting-semantics

In [17]:
!pip install jovian --upgrade --quiet

In [155]:
import jovian

In [157]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "ngo-minh-thang-nguyen/01-tensor-operations" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ml/ngo-minh-thang-nguyen/01-tensor-operations


'https://jovian.ml/ngo-minh-thang-nguyen/01-tensor-operations'